v2/tools/generator/internal/codegen/pipeline/stage.go

```diff
@@ -181,22 +184,61 @@ func (stage *Stage) Description() string {
 
 // Run is used to execute the action associated with this stage
 func (stage *Stage) Run(ctx context.Context, state *State) (*State, error) {
-	return stage.action(ctx, state)
+	if err := stage.checkPreconditions(state); err != nil {
```

Comment:
As discussed, there's a minor downside to having pre/post-requisites checkec during execution instead of up front, but I think that's well mitigated by having them always checked.